Maven Build Issue from Setting `Class-Path` in the jar manifest for grpc-netty-shaded

<!-- Please answer these questions before submitting a bug report. -->

### What version of gRPC-Java are you using?

1.41.0

### What is your environment?
<!-- operating system (Linux, Windows,...) and version, jdk version, etc. -->

JDK 16, Maven version 3.8.1, MacOS Catalina 10.15.7

### What did you expect to see?

a clean build of my project

### What did you see instead?

```
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/grpc-core-1.41.0.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/grpc-api-1.41.0.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/guava-30.1-android.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/error_prone_annotations-2.9.0.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/perfmark-api-0.23.0.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/failureaccess-1.0.1.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/jsr305-3.0.2.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/checker-compat-qual-2.5.5.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/j2objc-annotations-1.3.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/gson-2.8.6.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/annotations-4.1.1.4.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/animal-sniffer-annotations-1.19.jar": no such file or directory
[WARNING] [path] bad path element "/Users/mli/.m2/repository/io/grpc/grpc-netty-shaded/1.41.0/grpc-context-1.41.0.jar": no such file or directory
```

### Steps to reproduce the bug
I added some other dependencies to an existing project in my `pom.xml` such as 
```   
 <dependency>
      <groupId>org.grpcmock</groupId>
      <artifactId>grpcmock-junit5</artifactId>
      <version>${grpc-mock.version}</version>
</dependency>
```

<!-- Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs). -->

This issue has happened with other projects that we take a dependency on, see [here](https://github.com/launchdarkly/java-server-sdk/issues/240). Taking a look at the `MANIFEST.MF` of the grpc-netty-shaded jar, the `Class-Path` attribute is set which seems to be causing the bug.

```
Manifest-Version: 1.0
Implementation-Title: grpc-netty-shaded
Implementation-Version: 1.41.0
Class-Path: grpc-core-1.41.0.jar grpc-api-1.41.0.jar guava-30.1-androi
 d.jar error_prone_annotations-2.9.0.jar perfmark-api-0.23.0.jar failu
 reaccess-1.0.1.jar listenablefuture-9999.0-empty-to-avoid-conflict-wi
 th-guava.jar jsr305-3.0.2.jar checker-compat-qual-2.5.5.jar j2objc-an
 notations-1.3.jar gson-2.8.6.jar annotations-4.1.1.4.jar animal-sniff
 er-annotations-1.19.jar grpc-context-1.41.0.jar
```

I have a similar problem in Spring Boot application if run in Eclipse IDE. Here is my investigation of what is the cause of it:

The problem is about transitive dependency of `io.honeycomb` and classpath available for IDE. Not sure about IDEA, but in Eclipse all dependency classes are included to classpath.

Honeycomb adds a bunch of dependencies among which is also `grpc-netty-shaded-1.41.0.jar`:

```
...
|    ...
+--- io.honeycomb:honeycomb-opentelemetry-sdk:0.8.0
|    +--- io.opentelemetry:opentelemetry-exporter-logging:1.9.1
|    +--- io.opentelemetry:opentelemetry-exporter-otlp:1.9.1
|    +--- io.opentelemetry:opentelemetry-sdk-extension-resources:1.9.1
|    +--- io.grpc:grpc-netty-shaded:1.41.0
|    ...
...
```

Project have also implicit embedded tomcat dependency via `spring-boot-starter-web`, as well as explicit dependency of `tomcat-embed-jasper-9.0.56.jar`. Since we have `tomcat-embed-jasper` dependency a `org.apache.jasper.servlet.TldScanner` will be used to scan for Tag Library Descriptors during `org.apache.jasper.servlet.JasperInitializer.onStartup(Set<Class<?>>, ServletContext)` call.

Scanner will load available classpath entries resolving each of them. Library `grpc-netty-shaded-1.41.0.jar` is found successfully in local gradle cache:

```
file:/C:/Users/123/.gradle/caches/modules-2/files-2.1/io.grpc/grpc-netty-shaded/1.41.0/43fee785c101b3bfa3f7d916d8b47d5d736cbff2/grpc-netty-shaded-1.41.0.jar
```

Since it is a JAR file, an additional processing applied to check JAR file's `MANIFEST.MF`. It will extract all the entries of `Class-Path` of the manifest. Here is a list of all classpath entries of `grpc-netty-shaded-1.41.0.jar`:

```
Class-Path: grpc-core-1.44.0.jar grpc-api-1.44.0.jar guava-30.1.1-andr
 oid.jar error_prone_annotations-2.9.0.jar perfmark-api-0.23.0.jar fai
 lureaccess-1.0.1.jar listenablefuture-9999.0-empty-to-avoid-conflict-
 with-guava.jar jsr305-3.0.2.jar checker-compat-qual-2.5.5.jar j2objc-
 annotations-1.3.jar gson-2.8.9.jar annotations-4.1.1.4.jar animal-sni
 ffer-annotations-1.19.jar grpc-context-1.44.0.jar
```

Based on each of the entry an additional candidate will be added to the scan list. URL of the candidate is build in following way:

```java
    private void processManifest(Jar jar, boolean isWebapp,
            Deque<URL> classPathUrlsToProcess) throws IOException {
        ...
        Manifest manifest = jar.getManifest();
        if (manifest != null) {
            Attributes attributes = manifest.getMainAttributes();
            String classPathAttribute = attributes.getValue("Class-Path");
            ...
            String[] classPathEntries = classPathAttribute.split(" ");
            for (String classPathEntry : classPathEntries) {
                classPathEntry = classPathEntry.trim();
                ...
                URL jarURL = jar.getJarFileURL();
                URL classPathEntryURL;
                try {
                    URI jarURI = jarURL.toURI();
                    URI classPathEntryURI = jarURI.resolve(classPathEntry);
                    classPathEntryURL = classPathEntryURI.toURL();
                } catch (Exception e) {
                    if (log.isDebugEnabled()) {
                        log.debug(sm.getString("jarScan.invalidUri", jarURL), e);
                    }
                    continue;
                }
                classPathUrlsToProcess.add(classPathEntryURL);
            }
        }
    }
```

As a result it will create a list of additional entries similar to following:

```
file:/C:/Users/123/.gradle/caches/modules-2/files-2.1/io.grpc/grpc-netty-shaded/1.41.0/43fee785c101b3bfa3f7d916d8b47d5d736cbff2/grpc-core-1.41.0.jar
file:/C:/Users/123/.gradle/caches/modules-2/files-2.1/io.grpc/grpc-netty-shaded/1.41.0/43fee785c101b3bfa3f7d916d8b47d5d736cbff2/grpc-api-1.44.0.jar
file:/C:/Users/123/.gradle/caches/modules-2/files-2.1/io.grpc/grpc-netty-shaded/1.41.0/43fee785c101b3bfa3f7d916d8b47d5d736cbff2/guava-30.1.1-android.jar
...
```

Files above doesn't exists in a context of `grpc-netty-shaded`, they are all resided in their own contexts within the gradle cache. As a result I've got a lot of errors:

```
2022-01-28 18:28:24.475 [main] WARN  o.a.t.util.scan.StandardJarScanner   - Failed to scan [file:/C:/Users/123/.gradle/caches/modules-2/files-2.1/io.grpc/grpc-netty-shaded/1.41.0/43fee785c101b3bfa3f7d916d8b47d5d736cbff2/grpc-core-1.41.0.jar] from classloader hierarchy
java.io.IOException: java.lang.reflect.InvocationTargetException
	at org.apache.tomcat.util.compat.Jre9Compat.jarFileNewInstance(Jre9Compat.java:209)
    ...
Caused by: java.nio.file.NoSuchFileException: C:\Users\123\.gradle\caches\modules-2\files-2.1\io.grpc\grpc-netty-shaded\1.41.0\43fee785c101b3bfa3f7d916d8b47d5d736cbff2\grpc-core-1.41.0.jar
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
    ...
```

--- a/build.gradle
+++ b/build.gradle
@@ -369,6 +369,24 @@ subprojects {
         }
     }
 
+    plugins.withId("com.github.johnrengelman.shadow") {
+        tasks.named("shadowJar") {
+            // Do a dance to remove Class-Path. This needs to run after the doFirst() from the
+            // shadow plugin that adds Class-Path and before the core jar action. Using doFirst will
+            // have this run before the shadow plugin, and doLast will run after the core jar
+            // action. See #8606.
+            // The shadow plugin adds another doFirst when application is used for setting
+            // Main-Class. Ordering with it doesn't matter.
+            actions.add(plugins.hasPlugin("application") ? 2 : 1, new Action<Task>() {
+                @Override public void execute(Task task) {
+                    if (!task.manifest.attributes.remove("Class-Path")) {
+                        throw new AssertionError("Did not find Class-Path to remove from manifest")
+                    }
+                }
+            })
+        }
+    }
+
     plugins.withId("maven-publish") {
         publishing {
             publications {

